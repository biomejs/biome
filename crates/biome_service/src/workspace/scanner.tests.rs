use std::sync::Arc;

use biome_fs::{BiomePath, TemporaryFs};
use crossbeam::channel::bounded;
use tokio::sync::watch;

use crate::{
    Workspace, WorkspaceServer,
    workspace::{OpenProjectParams, ScanKind, ScanProjectFolderParams, ServiceDataNotification},
};

#[test]
#[cfg(any(target_os = "linux", target_os = "macos"))]
fn scanner_doesnt_show_errors_for_inaccessible_files() {
    use std::os::unix::fs::PermissionsExt;

    const FILE_CONTENT: &str = "import 'foo';";

    let mut fs = TemporaryFs::new("scanner_doesnt_show_errors_for_inaccessible_files");
    fs.create_file("a.js", FILE_CONTENT);

    let (watcher_tx, _) = bounded(0);
    let (service_data_tx, _) = watch::channel(ServiceDataNotification::Updated);
    let workspace =
        WorkspaceServer::new(Arc::new(fs.create_os()), watcher_tx, service_data_tx, None);

    let permissions = PermissionsExt::from_mode(0o000);
    std::fs::set_permissions(format!("{}/a.js", fs.cli_path()), permissions)
        .expect("Cannot set permissions");

    let result = workspace
        .open_project(OpenProjectParams {
            path: BiomePath::new(fs.cli_path()),
            open_uninitialized: true,
        })
        .unwrap();
    let project_key = result.project_key;

    let result = workspace
        .scan_project_folder(ScanProjectFolderParams {
            project_key,
            path: None,
            watch: false,
            force: false,
            scan_kind: ScanKind::Project,
            verbose: false,
        })
        .unwrap();

    assert!(result.diagnostics.is_empty());

    let result = workspace
        .scan_project_folder(ScanProjectFolderParams {
            project_key,
            path: None,
            watch: false,
            force: false,
            scan_kind: ScanKind::Project,
            verbose: true,
        })
        .unwrap();

    assert_eq!(result.diagnostics.len(), 1);
}
