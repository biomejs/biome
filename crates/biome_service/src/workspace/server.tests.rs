use biome_fs::MemoryFileSystem;

use super::*;

#[test]
fn commonjs_file_rejects_import_statement() {
    const FILE_CONTENT: &[u8] = b"import 'foo';";
    const MANIFEST_CONTENT: &[u8] = b"{ \"type\": \"commonjs\" }";

    let mut fs = MemoryFileSystem::default();
    fs.insert(Utf8PathBuf::from("/project/a.js"), FILE_CONTENT);
    fs.insert(Utf8PathBuf::from("/project/package.json"), MANIFEST_CONTENT);

    let workspace = WorkspaceServer::new(Box::new(fs));
    let project_key = workspace
        .open_project(OpenProjectParams {
            path: BiomePath::new("/"),
            open_uninitialized: true,
        })
        .unwrap();

    workspace
        .scan_project_folder(ScanProjectFolderParams {
            project_key,
            path: Some(BiomePath::new("/")),
        })
        .unwrap();

    match workspace.get_parse("/project/a.js".into()) {
        Ok(parse) => {
            insta::assert_debug_snapshot!(parse.diagnostics(), @r###"
            [
                ParseDiagnostic {
                    span: Some(
                        0..13,
                    ),
                    message: Illegal use of an import declaration outside of a module,
                    advice: ParserAdvice {
                        advice_list: [
                            Hint(
                                "not allowed inside scripts",
                            ),
                        ],
                    },
                },
            ]
            "###);
        }
        Err(error) => panic!("File not available: {error}"),
    }
}
