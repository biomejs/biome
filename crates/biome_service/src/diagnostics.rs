use crate::workspace::DocumentFileSource;
use biome_configuration::diagnostics::EditorConfigDiagnostic;
use biome_configuration::{CantLoadExtendFile, ConfigurationDiagnostic};
use biome_console::fmt::Bytes;
use biome_console::markup;
use biome_diagnostics::{
    category, Advices, Category, Diagnostic, DiagnosticTags, Location, LogCategory, Severity, Visit,
};
use biome_formatter::{FormatError, PrintError};
use biome_fs::{BiomePath, FileSystemDiagnostic};
use biome_grit_patterns::CompileError;
use biome_js_analyze::utils::rename::RenameError;
use biome_js_analyze::RuleError;
use serde::{Deserialize, Serialize};
use std::error::Error;
use std::ffi::OsStr;
use std::fmt;
use std::fmt::{Debug, Display, Formatter};
use std::process::{ExitCode, Termination};

/// Generic errors thrown during biome operations
#[derive(Deserialize, Diagnostic, Serialize)]
pub enum WorkspaceError {
    /// Can't export the report of the CLI into a file
    ReportNotSerializable(ReportNotSerializable),
    /// The project contains uncommitted changes
    DirtyWorkspace(DirtyWorkspace),
    /// The file does not exist in the [crate::Workspace]
    NotFound(NotFound),
    /// A file is not supported. It contains the language and path of the file
    /// Use this error if Biome is trying to process a file that Biome can't understand
    SourceFileNotSupported(SourceFileNotSupported),
    /// The formatter encountered an error while formatting the file
    FormatError(FormatError),
    /// The formatter encountered an error while formatting the file
    PrintError(PrintError),
    /// The file could not be formatted since it has syntax errors and `format_with_errors` is disabled
    FormatWithErrorsDisabled(FormatWithErrorsDisabled),
    /// The file could not be analyzed because a rule caused an error.
    RuleError(RuleError),
    /// Thrown when Biome can't read a generic directory
    CantReadDirectory(CantReadDirectory),
    /// Thrown when Biome can't read a generic file
    CantReadFile(CantReadFile),
    /// Error thrown when validating the configuration. Once deserialized, further checks have to be done.
    Configuration(ConfigurationDiagnostic),
    /// Error thrown when Biome cannot rename a symbol.
    RenameError(RenameError),
    /// Error emitted by the underlying transport layer for a remote Workspace
    TransportError(TransportError),
    /// Emitted when the file is ignored and should not be processed
    FileIgnored(FileIgnored),
    /// Emitted when a file could not be parsed because it's larger than the size limit
    FileTooLarge(FileTooLarge),
    /// Diagnostics emitted when querying the file system
    FileSystem(FileSystemDiagnostic),
    /// Raised when there's an issue around the VCS integration
    Vcs(VcsDiagnostic),
    /// Diagnostic raised when a file is protected
    ProtectedFile(ProtectedFile),
    /// Error when searching for a pattern
    SearchError(SearchError),
    /// Thrown when a `.editorconfig` file is found, but it can't be parsed.
    EditorConfigDiagnostic(EditorConfigDiagnostic),
}

impl WorkspaceError {
    pub fn format_with_errors_disabled() -> Self {
        Self::FormatWithErrorsDisabled(FormatWithErrorsDisabled)
    }

    pub fn cant_read_file(path: String) -> Self {
        Self::CantReadFile(CantReadFile { path })
    }

    pub fn not_found() -> Self {
        Self::NotFound(NotFound)
    }

    pub fn file_too_large(path: String, size: usize, limit: usize) -> Self {
        Self::FileTooLarge(FileTooLarge { path, size, limit })
    }

    pub fn file_ignored(path: String) -> Self {
        Self::FileIgnored(FileIgnored { path })
    }

    pub fn source_file_not_supported(
        language: DocumentFileSource,
        path: String,
        extension: Option<String>,
    ) -> Self {
        Self::SourceFileNotSupported(SourceFileNotSupported {
            file_source: language,
            path,
            extension,
        })
    }

    pub fn vcs_disabled() -> Self {
        Self::Vcs(VcsDiagnostic::DisabledVcs(DisabledVcs {}))
    }

    pub fn protected_file(file_path: impl Into<String>) -> Self {
        Self::ProtectedFile(ProtectedFile {
            file_path: file_path.into(),
            verbose_advice: ProtectedFileAdvice,
        })
    }
}

impl Error for WorkspaceError {}

impl Debug for WorkspaceError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        std::fmt::Display::fmt(self, f)
    }
}

impl Display for WorkspaceError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        Diagnostic::description(self, f)
    }
}

impl Termination for WorkspaceError {
    fn report(self) -> ExitCode {
        ExitCode::FAILURE
    }
}

impl From<FormatError> for WorkspaceError {
    fn from(err: FormatError) -> Self {
        Self::FormatError(err)
    }
}

impl From<TransportError> for WorkspaceError {
    fn from(err: TransportError) -> Self {
        Self::TransportError(err)
    }
}

impl From<PrintError> for WorkspaceError {
    fn from(err: PrintError) -> Self {
        Self::PrintError(err)
    }
}

impl From<FileSystemDiagnostic> for WorkspaceError {
    fn from(err: FileSystemDiagnostic) -> Self {
        Self::FileSystem(err)
    }
}

impl From<ConfigurationDiagnostic> for WorkspaceError {
    fn from(value: ConfigurationDiagnostic) -> Self {
        Self::Configuration(value)
    }
}

impl From<CantLoadExtendFile> for WorkspaceError {
    fn from(value: CantLoadExtendFile) -> Self {
        WorkspaceError::Configuration(ConfigurationDiagnostic::CantLoadExtendFile(value))
    }
}

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "internalError/fs",
    message = "Uncommitted changes in repository"
)]
pub struct DirtyWorkspace;

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "internalError/fs",
    message(
        message("The report can't be serialized, here's why: "{self.reason}),
        description = "The report can't be serialized, here's why: {reason}"
    )
)]
pub struct ReportNotSerializable {
    reason: String,
}

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "internalError/fs",
    message = "The file does not exist in the workspace.",
    tags(INTERNAL)
)]
pub struct NotFound;

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "format",
    message = "Code formatting aborted due to parsing errors. To format code with errors, enable the 'formatter.formatWithErrors' option."
)]
pub struct FormatWithErrorsDisabled;

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "internalError/fs",
    message(
        message("Biome couldn't read the following directory, maybe for permissions reasons or it doesn't exist: "{self.path}),
        description = "Biome couldn't read the following directory, maybe for permissions reasons or it doesn't exist: {path}"
    )
)]
pub struct CantReadDirectory {
    #[location(resource)]
    path: String,
}

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "internalError/fs",
    message(
        message("Biome couldn't read the following file, maybe for permissions reasons or it doesn't exist: "{self.path}),
        description = "Biome couldn't read the following file, maybe for permissions reasons or it doesn't exist: {path}"
    )
)]
pub struct CantReadFile {
    #[location(resource)]
    path: String,
}

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "internalError/fs",
    message(
        message("The file "{self.path}" was ignored."),
        description = "The file {path} was ignored."
    ),
    severity = Warning,
)]
pub struct FileIgnored {
    #[location(resource)]
    path: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FileTooLarge {
    path: String,
    size: usize,
    limit: usize,
}

impl Diagnostic for FileTooLarge {
    fn category(&self) -> Option<&'static Category> {
        Some(category!("internalError/fs"))
    }

    fn message(&self, fmt: &mut biome_console::fmt::Formatter<'_>) -> std::io::Result<()> {
        fmt.write_markup(
            markup!{
                "Size of "{self.path}" is "{Bytes(self.size)}" which exceeds configured maximum of "{Bytes(self.limit)}" for this project.
                The file size limit exists to prevent us inadvertently slowing down and loading large files that we shouldn't.
                Use the `files.maxSize` configuration to change the maximum size of files processed."
            }
        )
    }

    fn description(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
        write!(fmt,
               "Size of {} is {} which exceeds configured maximum of {} for this project.\n\
               The file size limit exists to prevent us inadvertently slowing down and loading large files that we shouldn't.\n\
               Use the `files.maxSize` configuration to change the maximum size of files processed.",
               self.path, Bytes(self.size), Bytes(self.limit)
        )
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SourceFileNotSupported {
    file_source: DocumentFileSource,
    path: String,
    extension: Option<String>,
}

impl Diagnostic for SourceFileNotSupported {
    fn category(&self) -> Option<&'static Category> {
        Some(category!("files/missingHandler"))
    }

    fn severity(&self) -> Severity {
        Severity::Warning
    }

    fn location(&self) -> Location<'_> {
        Location::builder().resource(&self.path).build()
    }

    fn message(&self, fmt: &mut biome_console::fmt::Formatter<'_>) -> std::io::Result<()> {
        if self.file_source != DocumentFileSource::Unknown {
            fmt.write_markup(markup! {
                "Biome doesn't support this feature for the language "{{&self.file_source}}
            })
        } else if let Some(ext) = self.extension.as_ref() {
            fmt.write_markup(markup! {
                "Biome could not determine the language for the file extension "{{ext}}
            })
        } else {
            fmt.write_markup(
                markup!{
                    "Biome could not determine the language for the file "{self.path}" because it doesn't have a clear extension"
                }
            )
        }
    }

    fn verbose_advices(&self, visitor: &mut dyn Visit) -> std::io::Result<()> {
        visitor.record_log(
            LogCategory::Info,
            &markup! {
                "If you want to turn off this diagnostic, consider using "<Emphasis>"--files-ignore-unknown"</Emphasis>" from the CLI, or "<Emphasis>"files.ignoreUnknown"</Emphasis>" from the configuration file."
            }
        )
    }
}

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
pub enum SearchError {
    /// An invalid pattern was given
    PatternCompilationError(CompileError),
    /// No pattern with the given ID
    InvalidPattern(InvalidPattern),
}

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "search",
    message(
        message("Invalid pattern -- this is a bug in Biome."),
        description = "If this problem persists, please report here: https://github.com/biomejs/biome/issues/"
    )
)]
pub struct InvalidPattern;

pub fn extension_error(path: &BiomePath) -> WorkspaceError {
    let file_source = DocumentFileSource::from_path(path);
    WorkspaceError::source_file_not_supported(
        file_source,
        path.clone().display().to_string(),
        path.clone()
            .extension()
            .and_then(OsStr::to_str)
            .map(|s| s.to_string()),
    )
}

#[derive(Debug, Serialize, Deserialize)]
/// Error emitted by the underlying transport layer for a remote Workspace
pub enum TransportError {
    /// Error emitted by the transport layer if the connection was lost due to an I/O error
    ChannelClosed,
    /// Error emitted by the transport layer if a request timed out
    Timeout,
    /// Error caused by a serialization or deserialization issue
    SerdeError(String),
    /// Generic error type for RPC errors that can't be deserialized into RomeError
    RPCError(String),
}

impl Display for TransportError {
    fn fmt(&self, fmt: &mut Formatter) -> fmt::Result {
        self.description(fmt)
    }
}

impl Diagnostic for TransportError {
    fn category(&self) -> Option<&'static Category> {
        Some(category!("internalError/io"))
    }

    fn severity(&self) -> Severity {
        Severity::Error
    }

    fn description(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
        match self {
            TransportError::SerdeError(err) => write!(fmt, "serialization error: {err}"),
            TransportError::ChannelClosed => fmt.write_str(
                "a request to the remote workspace failed because the connection was interrupted",
            ),
            TransportError::Timeout => {
                fmt.write_str("the request to the remote workspace timed out")
            }
            TransportError::RPCError(err) => fmt.write_str(err),
        }
    }

    fn message(&self, fmt: &mut biome_console::fmt::Formatter<'_>) -> std::io::Result<()> {
        match self {
            TransportError::SerdeError(err) => write!(fmt, "serialization error: {err}"),
            TransportError::ChannelClosed => fmt.write_str(
                "a request to the remote workspace failed because the connection was interrupted",
            ),
            TransportError::Timeout => {
                fmt.write_str("the request to the remote workspace timed out")
            }
            TransportError::RPCError(err) => fmt.write_str(err),
        }
    }
    fn tags(&self) -> DiagnosticTags {
        DiagnosticTags::INTERNAL
    }
}

#[derive(Debug, Deserialize, Diagnostic, Serialize)]
pub enum VcsDiagnostic {
    /// When the VCS folder couldn't be found
    NoVcsFolderFound(NoVcsFolderFound),
    /// VCS is disabled
    DisabledVcs(DisabledVcs),
}

impl From<VcsDiagnostic> for WorkspaceError {
    fn from(value: VcsDiagnostic) -> Self {
        Self::Vcs(value)
    }
}

impl From<CompileError> for WorkspaceError {
    fn from(value: CompileError) -> Self {
        Self::SearchError(SearchError::PatternCompilationError(value))
    }
}

#[derive(Debug, Diagnostic, Serialize, Deserialize)]
#[diagnostic(
    category = "internalError/fs",
    severity = Error,
    message(
        description = "Biome couldn't find the VCS folder at the following path: {path}",
        message("Biome couldn't find the VCS folder at the following path: "<Emphasis>{self.path}</Emphasis>),
    )
)]
pub struct NoVcsFolderFound {
    #[location(resource)]
    pub path: String,
}

#[derive(Debug, Diagnostic, Serialize, Deserialize)]
#[diagnostic(
    category = "internalError/fs",
    severity = Warning,
    message = "Biome couldn't determine a directory for the VCS integration. VCS integration will be disabled."
)]
pub struct DisabledVcs {}

#[derive(Debug, Serialize, Deserialize, Diagnostic)]
#[diagnostic(
    category = "project",
    severity = Information,
    message(
        message("The file "<Emphasis>{self.file_path}</Emphasis>" is protected because is handled by another tool. Biome won't process it."),
        description = "The file {file_path} is protected because is handled by another tool. Biome won't process it.",
    ),
    tags(VERBOSE)
)]
pub struct ProtectedFile {
    #[location(resource)]
    pub file_path: String,

    #[verbose_advice]
    pub verbose_advice: ProtectedFileAdvice,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProtectedFileAdvice;

impl Advices for ProtectedFileAdvice {
    fn record(&self, visitor: &mut dyn Visit) -> std::io::Result<()> {
        visitor.record_log(LogCategory::Info, &markup! { "You can hide this diagnostic by using "<Emphasis>"--diagnostic-level=warn"</Emphasis>" to increase the diagnostic level shown by CLI." })
    }
}

#[cfg(test)]
mod test {
    use crate::diagnostics::{
        CantReadDirectory, CantReadFile, DirtyWorkspace, FileIgnored, FileTooLarge, NotFound,
        SourceFileNotSupported,
    };
    use crate::file_handlers::DocumentFileSource;
    use crate::{TransportError, WorkspaceError};
    use biome_diagnostics::{print_diagnostic_to_string, DiagnosticExt, Error};
    use biome_formatter::FormatError;
    use biome_fs::BiomePath;
    use std::ffi::OsStr;

    fn snap_diagnostic(test_name: &str, diagnostic: Error) {
        let content = print_diagnostic_to_string(&diagnostic);

        insta::with_settings!({
            prepend_module_to_snapshot => false,
        }, {
            insta::assert_snapshot!(test_name, content);

        });
    }

    #[test]
    fn diagnostic_size() {
        assert_eq!(std::mem::size_of::<WorkspaceError>(), 96)
    }

    #[test]
    fn dirty_workspace() {
        snap_diagnostic(
            "dirty_workspace",
            WorkspaceError::DirtyWorkspace(DirtyWorkspace).into(),
        )
    }

    #[test]
    fn file_ignored() {
        snap_diagnostic(
            "file_ignored",
            WorkspaceError::FileIgnored(FileIgnored {
                path: "example.js".to_string(),
            })
            .with_file_path("example.js"),
        )
    }

    #[test]
    fn cant_read_directory() {
        snap_diagnostic(
            "cant_read_directory",
            WorkspaceError::CantReadDirectory(CantReadDirectory {
                path: "example/".to_string(),
            })
            .with_file_path("example/"),
        )
    }

    #[test]
    fn cant_read_file() {
        snap_diagnostic(
            "cant_read_file",
            WorkspaceError::CantReadFile(CantReadFile {
                path: "example.js".to_string(),
            })
            .with_file_path("example.js"),
        )
    }

    #[test]
    fn not_found() {
        snap_diagnostic(
            "not_found",
            WorkspaceError::NotFound(NotFound).with_file_path("not_found.js"),
        )
    }

    #[test]
    fn source_file_not_supported() {
        let path = BiomePath::new("not_supported.toml");
        snap_diagnostic(
            "source_file_not_supported",
            WorkspaceError::SourceFileNotSupported(SourceFileNotSupported {
                file_source: DocumentFileSource::Unknown,
                path: path.display().to_string(),
                extension: path
                    .extension()
                    .and_then(OsStr::to_str)
                    .map(|s| s.to_string()),
            })
            .with_file_path("not_supported.toml"),
        )
    }

    #[test]
    fn file_too_large() {
        snap_diagnostic(
            "file_too_large",
            WorkspaceError::FileTooLarge(FileTooLarge {
                path: "example.js".to_string(),
                limit: 100,
                size: 500,
            })
            .with_file_path("example.js"),
        )
    }

    #[test]
    fn transport_channel_closed() {
        snap_diagnostic(
            "transport_channel_closed",
            TransportError::ChannelClosed.into(),
        )
    }

    #[test]
    fn transport_timeout() {
        snap_diagnostic("transport_timeout", TransportError::Timeout.into())
    }

    #[test]
    fn transport_rpc_error() {
        snap_diagnostic(
            "transport_rpc_error",
            TransportError::RPCError("Some generic error".to_string()).into(),
        )
    }

    #[test]
    fn transport_serde_error() {
        snap_diagnostic(
            "transport_serde_error",
            TransportError::SerdeError("Some serialization/deserialization error".to_string())
                .into(),
        )
    }

    #[test]
    fn formatter_syntax_error() {
        snap_diagnostic(
            "formatter_syntax_error",
            WorkspaceError::FormatError(FormatError::SyntaxError).with_file_path("example.js"),
        )
    }
}
