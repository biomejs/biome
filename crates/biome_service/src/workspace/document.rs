use biome_parser::AnyParse;

use crate::diagnostics::FileTooLarge;

#[derive(Clone, Debug)]
pub(crate) struct Document {
    /// Document content.
    pub(crate) content: String,

    /// The version of the document.
    ///
    /// A version is only specified when the document is opened by a client,
    /// typically through the LSP. Documents that are only opened by the scanner
    /// do not have a version.
    pub(crate) version: Option<i32>,

    /// The index of where the original file source is saved.
    /// Use `WorkspaceServer#file_sources` to retrieve the file source that belongs to the document.
    pub(crate) file_source_index: usize,

    /// The result of the parser (syntax tree + diagnostics).
    /// Types explained:
    /// - `Option`: if the file can be read
    /// - `Result`: if the file is read, but the  file is too large
    /// - `AnyParse`: the result of the parsed file
    pub(crate) syntax: Option<Result<AnyParse, FileTooLarge>>,

    /// If `true`, this indicates the document has been opened by the scanner,
    /// and should be unloaded only when the project is unregistered.
    ///
    /// Note the file can still *also* be opened explicitly by a client such as
    /// the LSP Proxy. In that case `version` will be `Some` and
    /// `opened_by_scanner` will be `true`, and the document will only be
    /// unloaded when both are unset.
    pub(super) opened_by_scanner: bool,
}
