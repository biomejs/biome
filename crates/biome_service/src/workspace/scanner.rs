use biome_diagnostics::serde::Diagnostic;
use biome_diagnostics::{Diagnostic as _, Error, Severity};
use biome_fs::{BiomePath, PathInterner, TraversalContext, TraversalScope};
use camino::Utf8Path;
use crossbeam::channel::{unbounded, Receiver, Sender};
use std::collections::BTreeSet;
use std::panic::catch_unwind;
use std::sync::RwLock;
use std::thread;
use std::time::{Duration, Instant};
use tracing::instrument;

use crate::diagnostics::Panic;
use crate::projects::ProjectKey;
use crate::workspace::{DocumentFileSource, FileContent, OpenFileParams};
use crate::{Workspace, WorkspaceError};

use super::server::WorkspaceServer;

pub(crate) struct ScanResult {
    /// Diagnostics reported while scanning the project.
    pub diagnostics: Vec<Diagnostic>,

    /// Duration of the full scan.
    pub duration: Duration,
}

impl WorkspaceServer {
    #[instrument(level = "debug", skip(self))]
    pub(super) fn scan(
        &self,
        project_key: ProjectKey,
        folder: &Utf8Path,
    ) -> Result<ScanResult, WorkspaceError> {
        let (interner, _path_receiver) = PathInterner::new();
        let (diagnostics_sender, diagnostics_receiver) = unbounded();

        let collector = DiagnosticsCollector::new();

        let (duration, diagnostics) = thread::scope(|scope| {
            let handler = thread::Builder::new()
                .name("biome::scanner".to_string())
                .spawn_scoped(scope, || collector.run(diagnostics_receiver))
                .expect("failed to spawn scanner thread");

            // The traversal context is scoped to ensure all the channels it
            // contains are properly closed once scanning finishes.
            let duration = scan_folder(
                folder,
                ScanContext {
                    workspace: self,
                    project_key,
                    interner,
                    diagnostics_sender,
                    evaluated_paths: Default::default(),
                },
            );

            // Wait for the collector thread to finish.
            let diagnostics = handler.join().unwrap();

            (duration, diagnostics)
        });

        Ok(ScanResult {
            diagnostics,
            duration,
        })
    }
}

/// Initiates the filesystem traversal tasks from the provided path and runs it to completion.
///
/// Returns the duration of the process and the evaluated paths.
#[instrument(level = "debug", skip(ctx))]
fn scan_folder(folder: &Utf8Path, ctx: ScanContext) -> Duration {
    let start = Instant::now();
    let fs = ctx.workspace.fs();
    let ctx_ref = &ctx;
    fs.traversal(Box::new(move |scope: &dyn TraversalScope| {
        scope.evaluate(ctx_ref, folder.to_path_buf());
    }));

    let evaluated_paths = ctx.evaluated_paths();

    let mut configs = Vec::new();
    let mut manifests = Vec::new();
    let mut handleable_paths = Vec::with_capacity(evaluated_paths.len());
    let mut ignore_paths = Vec::new();
    for path in evaluated_paths {
        if path.is_config() {
            configs.push(path);
        } else if path.is_manifest() {
            manifests.push(path);
        } else if path.is_ignore() {
            ignore_paths.push(path);
        } else {
            handleable_paths.push(path);
        }
    }

    fs.traversal(Box::new(|scope: &dyn TraversalScope| {
        for path in &configs {
            scope.handle(ctx_ref, path.to_path_buf());
        }
    }));
    fs.traversal(Box::new(|scope: &dyn TraversalScope| {
        for path in &manifests {
            scope.handle(ctx_ref, path.to_path_buf());
        }
    }));

    let mut paths = configs;
    paths.append(&mut manifests);
    ctx.workspace.update_project_layout_for_paths(&paths);
    let result = ctx
        .workspace
        .update_project_ignore_files(ctx.project_key, &ignore_paths);

    if let Err(error) = result {
        ctx.send_diagnostic(error);
    }

    fs.traversal(Box::new(|scope: &dyn TraversalScope| {
        for path in &handleable_paths {
            scope.handle(ctx_ref, path.to_path_buf());
        }
    }));

    ctx.workspace
        .update_dependency_graph_for_paths(&handleable_paths);

    start.elapsed()
}

struct DiagnosticsCollector {
    /// The minimum level of diagnostic we should collect.
    diagnostic_level: Severity,

    /// Whether we should collect verbose diagnostics.
    verbose: bool,
}

impl DiagnosticsCollector {
    fn new() -> Self {
        Self {
            diagnostic_level: Severity::Hint,
            verbose: false,
        }
    }

    /// Checks whether the given `diagnostic` should be collected or not.
    fn should_collect(&self, diagnostic: &Diagnostic) -> bool {
        diagnostic.severity() >= self.diagnostic_level
            && (self.verbose || !diagnostic.tags().is_verbose())
    }

    fn run(&self, receiver: Receiver<Diagnostic>) -> Vec<Diagnostic> {
        let mut diagnostics = Vec::new();

        while let Ok(diagnostic) = receiver.recv() {
            if self.should_collect(&diagnostic) {
                diagnostics.push(diagnostic);
            }
        }

        diagnostics
    }
}

/// Context object shared between directory traversal tasks.
pub(crate) struct ScanContext<'app> {
    /// [Workspace] instance.
    pub(crate) workspace: &'app WorkspaceServer,

    /// Key of the project within which this scanner is active.
    project_key: ProjectKey,

    /// File paths interner cache used by the filesystem traversal.
    interner: PathInterner,

    /// Channel for reporting diagnostics during scanning.
    pub(crate) diagnostics_sender: Sender<Diagnostic>,

    /// List of paths that should be processed.
    pub(crate) evaluated_paths: RwLock<BTreeSet<BiomePath>>,
}

impl ScanContext<'_> {
    /// Send a message to the display thread
    pub(crate) fn send_diagnostic(&self, diagnostic: impl Into<Diagnostic>) {
        self.diagnostics_sender.send(diagnostic.into()).ok();
    }
}

impl TraversalContext for ScanContext<'_> {
    fn interner(&self) -> &PathInterner {
        &self.interner
    }

    fn evaluated_paths(&self) -> BTreeSet<BiomePath> {
        self.evaluated_paths.read().unwrap().clone()
    }

    fn push_diagnostic(&self, error: Error) {
        self.send_diagnostic(Diagnostic::new(error));
    }

    fn can_handle(&self, path: &BiomePath) -> bool {
        path.is_dir() || DocumentFileSource::try_from_path(path).is_ok() || path.is_ignore()
    }

    fn handle_path(&self, path: BiomePath) {
        open_file(self, &path)
    }

    fn store_path(&self, path: BiomePath) {
        self.evaluated_paths
            .write()
            .unwrap()
            .insert(BiomePath::new(path.as_path()));
    }
}

/// Instructs the workspace to open a single file and submits diagnostics in
/// case of an error.
///
/// The call to the workspace method is also wrapped in a [catch_unwind] block
/// so panics are caught, and diagnostics are submitted in case of panic too.
fn open_file(ctx: &ScanContext, path: &BiomePath) {
    match catch_unwind(move || {
        ctx.workspace.open_file_by_scanner(OpenFileParams {
            project_key: ctx.project_key,
            path: path.clone(),
            content: FileContent::FromServer,
            document_file_source: None,
            version: 0,
            persist_node_cache: false,
        })
    }) {
        Ok(Ok(())) => {}
        Ok(Err(err)) => {
            ctx.send_diagnostic(err);
        }
        Err(err) => {
            let error = match err.downcast::<String>() {
                Ok(description) => Panic::with_file_and_message(path, *description),
                Err(err) => match err.downcast::<&'static str>() {
                    Ok(description) => Panic::with_file_and_message(path, *description),
                    Err(_) => Panic::with_file(path),
                },
            };

            ctx.send_diagnostic(error);
        }
    }
}
