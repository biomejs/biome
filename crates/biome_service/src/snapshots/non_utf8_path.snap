---
source: crates/biome_service/src/diagnostics.rs
expression: content
---
path.js internalError/fs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Biome does not support non-UTF8 characters in path: path.js
